net/tstun/linkattrs_linux.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tstun
+
+import (
+	"github.com/mdlayher/genetlink"
+	"github.com/mdlayher/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/tun"
+)
+
+// linkSpeedUnknown is SPEED_UNKNOWN from linux/ethtool.h,
+// which is appropriate for a virtual interface.
+const linkSpeedUnknown = ^uint32(0)
+
+// setLinkSpeed sets the advertised link speed of the TUN interface.
+func setLinkSpeed(iface tun.Device, mbps uint32) error {
+	name, err := iface.Name()
+	if err != nil {
+		return err
+	}
+
+	conn, err := genetlink.Dial(&netlink.Config{Strict: true})
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+
+	f, err := conn.GetFamily(unix.ETHTOOL_GENL_NAME)
+	if err != nil {
+		return err
+	}
+
+	ae := netlink.NewAttributeEncoder()
+	ae.Nested(unix.ETHTOOL_A_LINKMODES_HEADER, func(nae *netlink.AttributeEncoder) error {
+		nae.String(unix.ETHTOOL_A_HEADER_DEV_NAME, name)
+		return nil
+	})
+	ae.Uint32(unix.ETHTOOL_A_LINKMODES_SPEED, mbps)
+
+	b, err := ae.Encode()
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.Execute(
+		genetlink.Message{
+			Header: genetlink.Header{
+				Command: unix.ETHTOOL_MSG_LINKMODES_SET,
```

Comment:
AFAIK all of the set operations require CAP_NET_ADMIN but you probably need that in other areas in Tailscale anyway?

Reply:
Except in userspace-networking mode AFAIU, but this code won't run in that mode (and failure is OK in this path anyway).